core/pageProcesser/cs_page.c

```diff
@@ -0,0 +1,84 @@
+#include <string.h>
+#include "pageProcesser.h"
+#include "cs_page.h"
+
+void clear_page(cs_page *p);
+void destroy_page(cs_page *p);
+int new_page(cs_page *p, unsigned int capacity);
+int set_page(cs_page *p, char* context, unsigned int length);
+
+/*
+  clear_page: fill the data with all zeros.
+*/
+
+void clear_page(cs_page *p) {
+  if(p == NULL)
+    return;
+  if(p->data == NULL)
+    return;
+  memset(p->data, 0, p->capacity);
+  p->used = 0;
+  p->file_type = FileTypeErr;
+}
+
+/*
+  destroy_page: recycle the memory. Optionally before new_page or set_page.
+*/
+
+void destroy_page(cs_page *p) {
+  if(p == NULL)
+    return;
+  if(p->data != NULL)
+    free(p->data);
+  memset(p, 0, sizeof(cs_page));
+}
+
+/*
+  new_page: pre-allocate the memory before using, optionally.
+*/
+
+int new_page(cs_page *p, unsigned int capacity) {
+  if(p == NULL)
+    return 0x1;
+  if(capacity == 0)
+    return 0x10;
+  void* buf = malloc(capacity);
+  if(buf == NULL)
+    return 0x100;
+  p->data = buf;
+  p->capacity = capacity;
+  memset(p->data, 0, p->capacity);
+  /* set p->file_type and p->used to zeros */
+  memset(&p->file_type, 0, sizeof(cs_page) - &0->file_type);
```

Comment:
Absolutely not. &0->filye_type is a magic for calculating the length of 0 to file_type.